Feature/premakedeps

Ohjurot opened the pull request:
Changelog: Feature: Adding preliminary (non documented, dev-only) support for premake5 deps (PremakeDeps).
Docs: Omit

Fixes #13275 (Based on PR #13350)

**Features**
- Implements lua file generation for premake consumption (New feature for 2.0)
- Supports multi build with changing architecture and configuration (New for the premake tool)
- Generates a lua file with variables (includedirs, libdirs, ...) per configuration (Based on existing implementation, adapted for multi build and new architecture)
- Generates a lua file with functions (Just build/compiler setup, linker setup and both) for automatic premake project setup per package and profile/config (Breaks compatibility with conan1 but is way more flexible)
- Generates a lua file with functions (see previous) for automatic premake project setup per package
- Generates a global `conandeps.premake5.lua` file with functions (see previous) for automatic premake project setup. 
- Has fallback capabilities to `release` when no fitting conan configuration is available

**Status**
WIP-Status: Code works and has been covered with a basic unit test. Documentation is still required.

**Checklist**
- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] ~~I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.~~ Delayed until full premake5 support is implemented.

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

Ohjurot committed:
Step 1: Generate premake files per package config

Ohjurot committed:
Step 2: Produce files per package and global

Ohjurot committed:
Documentation & Fixed typo

Ohjurot committed:
Make the style better

Ohjurot committed:
- Added unittest
- Fixed bugs that only occur in "fresh" deployments
- Fixed bugs that only occur in edge cases

Ohjurot committed:
Applied first feedback and fixed non consistent tabs.

Ohjurot commented:
@memsharded In #13350 you have mentioned "As long as we don't document it, it should be very doable". So, shall I wait with writing the documentation until I implement the full premake toolchain? So that you can merge this PR and get the `PremakeDeps` support in as soon as possible, but without officially stating it trough the documentation?

memsharded commented:
> @memsharded In https://github.com/conan-io/conan/pull/13350 you have mentioned "As long as we don't document it, it should be very doable". So, shall I wait with writing the documentation until I implement the full premake toolchain? So that you can merge this PR and get the PremakeDeps support in as soon as possible, but without officially stating it trough the documentation?

Yes, exactly, it is possible to merge it to allow easier preliminary testing for early adopters, contributors and maintainers, without being too constrained to change things. I'd wait until we have the basic pieces working together and a bit of our own testing.

memsharded commented on conan/tools/premake/premakedeps.py:
```diff
@@ -37,74 +109,217 @@ def __init__(self, deps_cpp_info):
+        """
+        - Uses filters that serve the same purpose than ``_premake_filtered()``
+        - This call will create an inverse filter on configurations. It will only apply if non of the 
+          ``configurations`` from this function call is present in premake. This is the fallback when a premake
```
I am not sure about this functionality, other build systems do not implement it, but need the explicit settings to match (it is possible to define different configs with ``-s &:build_type=Debug -s build_type=Release``, but in any case, the config of the dependency is well defined, no fallback (that might cause issues in OS like Windows). Maybe we can start with something simpler without it?

Ohjurot replied:
I have added this feature because I faced an issue several times in the past. 

For example if I create a configuration `Profile` in premake and try to apply the automatic `conan_setup()` function(s) it will end up with not applying the dependencys at all. With the implemented functionality it will automatically apply the `release` dependencys to all unknown configurations. This is an edge case that i have faced some times. 

I would prevere keeping it because I'm facing that edge case in several project. Not keeping that functionality would require me to manually use the lua variables and not be able to use the automatically generate functions. However if this functionality is not wanted I can remove it...

Ohjurot replied:
With my latest commit this should no longer be relevant.

memsharded commented on conan/tools/premake/premakedeps.py:
```diff
@@ -37,74 +109,217 @@ def __init__(self, deps_cpp_info):
+                fallback_configuration = configurations[-1]
+            
+            # Emit package premake file
+            pkg_files.append(PREMAKE_PKG_FILE.format(pkgname=dep_name))
+            self._output_lua_file(pkg_files[-1], [
```
```suggestion
            pkg_filename = PREMAKE_PKG_FILE.format(pkgname=dep_name)
            pkg_files.append(pkg_filename)
            self._output_lua_file(pkg_filename, [
```

Just a small detail, we typically prefer this more explicit linear style

memsharded commented on conan/tools/premake/premakedeps.py:
```diff
@@ -1,10 +1,82 @@
+function conan_setup_link_{pkgname}{config}()
+{conf_consume_link}
+end
+function conan_setup_{pkgname}{config}()
```
No idea if this is even possible, but maybe it would be more idiomatic to have functions like ``conan_setup(pkg_name, config)`` as parameters? and then it will be called ``conan_setup(mypkg, Release)`` or something like that, and that would avoid a lot of generated code?

Ohjurot replied:
I'm going to take a look at how this is possible in lua. I will respond to this as soon as I have more information.

Ohjurot replied:
I have found a way to do it like suggested. It required rewriting the way the lua variables are generated and consumed. It also made the python code around 50 lines shorter. 

 Doing it like suggested is also removing the need for the `_premake_filtered_fallback` (Python) feature. This can now be replaced by a call to `conan_setup(conf)` (Lua) over the automatic `conan_setup()`(Lua).

The lua function calls work in cascades:
- `conan_setup()` will call all possible
- `conan_setup(conf)` that call 
- `conan_setup(conf, pkg)` for all packages

I will do a few more tests with my own codebase to see if the implementation holds up and then do the commit.

memsharded reviewed (commented):
Some minor comments and doubts, please have a look.

Then, would you say this is ready to move forward, merge (quietly, no docs yet) it, and keep working on the other pieces?

Ohjurot committed:
Update conan/tools/premake/premakedeps.py

Added feedback for appending and referencing by explicit filename.

Co-authored-by: James <memsharded@gmail.com>

Ohjurot commented:
> Some minor comments and doubts, please have a look.
> 
> Then, would you say this is ready to move forward, merge (quietly, no docs yet) it, and keep working on the other pieces?

Yes, as soon as we have discussed the two open points this should be ready for a silent merge.

Ohjurot committed:
Included multiple feedback:
- Uncommon feature: "_premake_filtered_fallback"
- Better consumption with a function in style of "conan_setup(conf, pkg)"
By:
- Change lua variable to be stored in a table
- Changed consumption of lua variables

Ohjurot commented:
@memsharded can you please give me an update

memsharded commented:
Sorry, I have been buried by a lot of other things. Just assigned this for next 2.0.4 so we can prioritize it.

memsharded commented:
Merged! It will be in next 2.0.4
